Templates vignette (#122)
GSK-Biostatistics/tfrmt: vignettes/templates.Rmd

---
title: "Creating Template tfrmts"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Creating Template tfrmts}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(tfrmt)
library(dplyr)
library(gt)
library(tidyr)
```

A core design element of `tfrmt` is to offer the ability to layer tfrmts together. This ability provides an opportunity for organizations to build template table formats that can be shared, improved, and reused across multiple projects. Through layering the users can customize the templates to create a specific table. 

A template is a function that given a set of inputs (or none) create a standard `tfrmt`. We will go through their applications below.  

> 

Author: ```suggestion
We will create an Adverse Events table template in this vignette to demonstrate the ideas behind developing a template tfrmt.
```

> The creation of a template is based on the standard development of a `tfrmt`, but wrapping its creation into a function. using the `layer_tfrmt` function. We will go through their applications below.  

Reviewer: ```suggestion
The creation of a template involves build a standard `tfrmt`, then wrapping it into a function using the `layer_tfrmt` function. We will go through this application below.  
```

## Creating A Standard tfrmt

> A template is based on creating a standard `tfrmt`, and then writing a function around it to provide the ability to layer multiple templates. 

Author: ```suggestion
```

> A template is based around creating a standard `tfrmt`, and then writing a function around it to provide the ability to layer multiple templates. 

Reviewer: ```suggestion
A template is based on creating a standard `tfrmt`, and then writing a function around it to provide the ability to layer multiple templates. 
```

> We will create an Adverse Events table template in this example. This requires there to be some standards that can be expected that allows the construction of the basic `tfrmt`:

Reviewer: It might be better to move the  first sentence "we will create an AE table template.." to before the two sections to set the example up for the reader

Author: ```suggestion
The first step to creating an Adverse Events table template is to determine the constants in the format. This requires there to be some standards that can be expected that allows the construction of the basic `tfrmt`:
```

  
```{r basic-template-tfrmt}

ae_tfrmt_template <- tfrmt(
    group = AEBODSYS,
    label = AETERM,
    param = param,
    column = treatment,
    value = value,
  body_plan = body_plan(
    ## All entries where the param column is `AEs` (representing total counts)
    frmt_structure(
      group_val = ".default",
      label_val = ".default",
      AEs = frmt("[XXX]")
      ),
    
    ## Combine entries where param column is `n` and `pct` to create a cell for
    ## that population
    frmt_structure(
      group_val = ".default",
      label_val = ".default",
      frmt_combine(
        "{n} {pct}",
        n = frmt("XXX"),
        pct = frmt_when(
          "==100" ~ "",
          "==0" ~ "",
          TRUE ~ frmt("(xx.x %)")
          )
        )
      ),
    
    ## All entries where param column is `pval`, format conditionally.
    ## When the value is missing, replace the NA with "--".
    frmt_structure(
       group_val = ".default", 
       label_val = ".default",
       pval = frmt_when(
         ">0.99" ~ ">0.99",
         "<0.001" ~ "<0.001",
         TRUE ~ frmt("x.xxx", missing = "--")
         )
    )
  )
)

```

## Functionalizing The Template

> This is where we turn to using functions to define a template. This allows us to have as many layers as we'd like. For example, we can have an organization-level template to be used across all tables, a domain-level template to be used across all tables in a given domain, and a project-level template to be used for study-specific tables. 

Reviewer: ```suggestion
Now that we have our tfrmt, we can wrap it into a function to form a template. This allows us to have as many layers as we'd like. For example, we can have an organization-level template to be used across all tables, a domain-level template to be used across all tables in a given domain, and a project-level template to be used for study-specific tables. 
```

> `tfrmt` offers the function `layer_tfrmt()`, which provides the ability for layering `tfrmt` together. The first two arguments are the `tfrmt` objects to be layered. By default the body_plans of the tfrmt are joined together.

Reviewer: ```suggestion
`tfrmt` offers the function `layer_tfrmt()`, which provides the ability for layering `tfrmt` together. The first two arguments are the `tfrmt` objects to be layered. By default the body_plans of the tfrmt are joined together. We will leverage this function in the creation of our template.
```

        "Xanomeline High Dose (N=84)",
        `n (%)` = `Xanomeline High Dose-n_pct` ,
        `[AEs]` = `Xanomeline High Dose-AEs`
      ),
      span_structure(
        "Xanomeline Low Dose (N=84)",
        `n (%)` = `Xanomeline Low Dose-n_pct` ,
        `[AEs]` = `Xanomeline Low Dose-AEs`
      ),
      span_structure("Placebo (N=86)",
                     `n (%)` = `Placebo-n_pct` ,
                     `[AEs]` = `Placebo-AEs`),
      span_structure(
        "Fisher's Exact p-values",
        `Placebo vs. Low Dose` = `fisher_pval-p_low` ,
        `Placebo vs. High Dose` = `fisher_pval-p_high`
      )
    )
  )

```

See how this results in the same table as above:

```{r tfrmt-template-table}

## filter to keep only AEs with >10% prevalence in the High Dose group
ae_data2 <- ae_data %>% 
  group_by(AEBODSYS, AETERM) %>% 
  mutate(pct_high = value[col2=="Xanomeline High Dose" & param=="pct"]) %>% 
  ungroup %>% 
  filter(pct_high >10) %>% 
  select(-pct_high) %>% 
  unite("treatment", c(col2, col1), sep = "-")

study_ae_tfrmt_multi_layer %>% 
  print_to_gt(ae_data2) %>% 
  tab_options(
    container.width = 1000
  )

```

## Alternatives to Full Templates 

> In addition to defining full tables, frmts can also be defined at an organizational level. The benefit of this is that teams can use these frmt's in their tfrmt directly and can be sure their table will comply with the value presentation expected of their organization.

Reviewer: ```suggestion
In addition to defining full tfrmt templates, users can also create reusable frmt templates. The benefit of this is that teams can use these frmt's in their tfrmt directly and can be sure their table will comply with the value presentation expected of their organization.
```

      
      frmt_structure(
         group_val = ".default", 
         label_val = ".default",
         pval = pval_frmt
      )
    ),
    
    ## remove order columns from final table
    col_plan = col_plan(
      -ord1,-ord2,
      span_structure(
        "Xanomeline High Dose (N=84)",
        `n (%)` = `Xanomeline High Dose-n_pct` ,
        `[AEs]` = `Xanomeline High Dose-AEs`
      ),
      span_structure(
        "Xanomeline Low Dose (N=84)",
        `n (%)` = `Xanomeline Low Dose-n_pct` ,
        `[AEs]` = `Xanomeline Low Dose-AEs`
      ),
      span_structure("Placebo (N=86)",
                     `n (%)` = `Placebo-n_pct` ,
                     `[AEs]` = `Placebo-AEs`),
      span_structure(
        "Fisher's Exact p-values",
        `Placebo vs. Low Dose` = `fisher_pval-p_low` ,
        `Placebo vs. High Dose` = `fisher_pval-p_high`
      )
    )
    
  ) %>% 
  print_to_gt(ae_data2) %>% 
  tab_options(
    container.width = 1000
  )

```

## Best Practices

Here we list some of the ideas about best practices when it comes to defining tfrmt templates for use across organization. This comes from experience writing and using `tfrmt` and has the potential to change. 

> Creating a template tfrmt is as much about the table you are trying to standardize as what should be considered to change across tables. These functions should serve to provide a way to interact with and create tfrmt's for standard tables in such a way that customization for a specific table's `tfrmt` is simplified. However, as with normal function development, this should be balanced with limiting the number of arguments necessary to pass and get a functional tfrmt. For example, `tfrmt_sigdig` accepts a minimal set of arguments, including a data.frame or tibble, to define a complex body_plan. Simplifying the way users may enter this complex set of instructions without having to manually type it is part of its strength.

Reviewer: ```suggestion
When creating a template tfrmt, it is helpful to capture standard aspects of the data/display, rather than aspects that are likely to change across tables or studies. These functions should serve to provide a way to interact with and create tfrmt's for standard tables in such a way that customization for a specific table's `tfrmt` is simplified. However, as with normal function development, this should be balanced with limiting the number of arguments necessary to pass and get a functional tfrmt. For example, `tfrmt_sigdig` accepts a minimal set of arguments, including a data.frame or tibble, to define a complex body_plan. Simplifying the way users may enter this complex set of instructions without having to manually type it is part of its strength.
```
